File: benches/fplll_compare.cpp
// In-process fplll side of the lattica comparison benchmark.

#include <algorithm>
#include <chrono>
#include <cstdint>
#include <cstdlib>
#include <iomanip>
#include <iostream>
#include <stdexcept>
#include <tuple>
#include <vector>

#include <fplll/fplll.h>

using fplll::closest_vector;
using fplll::CVPM_FAST;
using fplll::CVPM_PROVED;
using fplll::FT_DEFAULT;
using fplll::LLL_DEF_ETA;
using fplll::LLL_DEFAULT;
using fplll::lll_reduction;
using fplll::LM_WRAPPER;
using fplll::RED_SUCCESS;
using fplll::Z_NR;
using fplll::ZZ_mat;

namespace {

constexpr int dimensions[] = {8, 16, 24};
constexpr std::size_t lll_cases = 16;
constexpr std::size_t targets_per_dimension = 128;
constexpr std::size_t samples = 11;
constexpr std::int64_t cvp_scale = 1'009;
volatile int status_sink = 0;

struct Rng {
  std::uint64_t state;

  std::uint64_t next() {
    state ^= state << 13;
    state ^= state >> 7;
    state ^= state << 17;
    return state;
  }

  std::size_t index(std::size_t bound) {
    return static_cast<std::size_t>(next() % bound);
  }

  std::int64_t signed_value(std::int64_t bound) {
    const auto width = static_cast<std::uint64_t>(2 * bound + 1);
    return static_cast<std::int64_t>(next() % width) - bound;
  }
};

using Rows = std::vector<std::int64_t>;

Rows canonical_basis(std::size_t dimension) {
  Rows basis(dimension * dimension);
  for (std::size_t row = 0; row < dimension; ++row) {
    basis[row * dimension + row] = 2;
    if (row > 0) {
      basis[row * dimension + row - 1] = 1;
    }
    if (row > 1) {
      basis[row * dimension + row - 2] = -1;
    }
  }
  return basis;
}

Rows cvp_basis(std::size_t dimension) {
  Rows basis(dimension * dimension);
  for (std::size_t row = 0; row < dimension; ++row) {
    basis[row * dimension + row] = 2;
    if (row + 1 < dimension) {
      basis[row * dimension + row + 1] = 1;
    }
  }
  return basis;
}

Rows skew_basis(std::size_t dimension, std::size_t benchmark_case) {
  auto basis = canonical_basis(dimension);
  Rng rng{0xd1b54a32d192ed03ULL ^ static_cast<std::uint64_t>(dimension) ^
          (static_cast<std::uint64_t>(benchmark_case) << 32)};
  for (std::size_t step = 0; step < 2 * dimension; ++step) {
    const auto destination = rng.index(dimension);
    auto source = rng.index(dimension - 1);
    if (source >= destination) {
      ++source;
    }
    const std::int64_t sign = (rng.next() & 1) == 0 ? -1 : 1;
    const auto destination_start = destination * dimension;
    const auto source_start = source * dimension;
    bool acceptable = true;
    for (std::size_t column = 0; column < dimension; ++column) {
      const auto candidate = basis[destination_start + column] +
                             sign * basis[source_start + column];
      acceptable = acceptable && std::abs(candidate) <= 256;
    }
    if (acceptable) {
      for (std::size_t column = 0; column < dimension; ++column) {
        basis[destination_start + column] +=
            sign * basis[source_start + column];
      }
    }
  }
  return basis;
}

ZZ_mat<mpz_t> to_matrix(const Rows &rows, std::size_t dimension) {
  ZZ_mat<mpz_t> matrix(static_cast<int>(dimension),
                       static_cast<int>(dimension));
  for (std::size_t row = 0; row < dimension; ++row) {
    for (std::size_t column = 0; column < dimension; ++column) {
      matrix[static_cast<int>(row)][static_cast<int>(column)] =
          rows[row * dimension + column];
    }
  }
  return matrix;
}

std::vector<Rows> make_targets(const Rows &basis, std::size_t dimension) {
  Rng rng{0xa0761d6478bd642fULL ^ static_cast<std::uint64_t>(dimension)};
  std::vector<Rows> targets;
  targets.reserve(targets_per_dimension);
  for (std::size_t target_index = 0; target_index < targets_per_dimension;
       ++target_index) {
    std::vector<std::int64_t> coordinates(dimension);
    for (auto &coordinate : coordinates) {
      coordinate = rng.signed_value(8);
    }
    Rows target(dimension);
    for (std::size_t row = 0; row < dimension; ++row) {
      for (std::size_t column = 0; column < dimension; ++column) {
        target[column] +=
            coordinates[row] * basis[row * dimension + column] * cvp_scale;
      }
    }
    for (auto &entry : target) {
      entry += rng.signed_value(1'000);
    }
    targets.push_back(std::move(target));
  }
  return targets;
}

std::vector<Z_NR<mpz_t>> to_vector(const Rows &values) {
  std::vector<Z_NR<mpz_t>> result(values.size());
  for (std::size_t index = 0; index < values.size(); ++index) {
    result[index] = values[index];
  }
  return result;
}

int reduce(ZZ_mat<mpz_t> &basis) {
  return lll_reduction(basis, 0.99, LLL_DEF_ETA, LM_WRAPPER, FT_DEFAULT, 0,
                       LLL_DEFAULT);
}

double median_ns(std::vector<std::chrono::nanoseconds> timings,
                 std::size_t operations) {
  std::sort(timings.begin(), timings.end());
  return static_cast<double>(timings[timings.size() / 2].count()) /
         static_cast<double>(operations);
}

std::pair<double, std::int64_t> benchmark_lll(std::size_t dimension) {
  std::vector<Rows> rows;
  std::vector<ZZ_mat<mpz_t>> matrices;
  rows.reserve(lll_cases);
  matrices.reserve(lll_cases);
  std::int64_t checksum = 0;
  std::size_t flat_index = 0;
  for (std::size_t benchmark_case = 0; benchmark_case < lll_cases;
       ++benchmark_case) {
    rows.push_back(skew_basis(dimension, benchmark_case));
    for (const auto entry : rows.back()) {
      ++flat_index;
      checksum += static_cast<std::int64_t>(flat_index) * entry;
    }
    matrices.push_back(to_matrix(rows.back(), dimension));
  }

  for (const auto &source : matrices) {
    auto basis = source;
    const int status = reduce(basis);
    if (status != RED_SUCCESS) {
      throw std::runtime_error("fplll LLL warm-up failed");
    }
    status_sink = status;
  }

  std::vector<std::chrono::nanoseconds> timings;
  timings.reserve(samples);
  for (std::size_t sample = 0; sample < samples; ++sample) {
    const auto start = std::chrono::steady_clock::now();
    for (const auto &source : matrices) {
      auto basis = source;
      const int status = reduce(basis);
      if (status != RED_SUCCESS) {
        throw std::runtime_error("fplll LLL failed");
      }
      status_sink = status;
    }
    timings.push_back(std::chrono::duration_cast<std::chrono::nanoseconds>(
        std::chrono::steady_clock::now() - start));
  }
  return {median_ns(std::move(timings), lll_cases), checksum};
}

std::tuple<std::int64_t, std::int64_t, std::int64_t>
fingerprints(const ZZ_mat<mpz_t> &basis,
             const std::vector<std::vector<Z_NR<mpz_t>>> &targets,
             const std::vector<std::vector<Z_NR<mpz_t>>> &coordinates) {
  const auto dimension = static_cast<std::size_t>(basis.get_cols());
  std::int64_t target_checksum = 0;
  std::int64_t point_checksum = 0;
  std::int64_t distance_checksum = 0;
  for (std::size_t target_index = 0; target_index < coordinates.size();
       ++target_index) {
    std::int64_t distance = 0;
    for (std::size_t column = 0; column < dimension; ++column) {
      Z_NR<mpz_t> ambient;
      for (std::size_t row = 0; row < dimension; ++row) {
        ambient.addmul(coordinates[target_index][row],
                       basis[static_cast<int>(row)][static_cast<int>(column)]);
      }
      const auto ambient_scaled = ambient.get_si();
      const auto weight =
          static_cast<std::int64_t>(1 + target_index * dimension + column);
      point_checksum += weight * (ambient_scaled / cvp_scale);
      target_checksum += weight * targets[target_index][column].get_si();
      const auto residual =
          targets[target_index][column].get_si() - ambient_scaled;
      distance += residual * residual;
    }
    distance_checksum += static_cast<std::int64_t>(target_index + 1) * distance;
  }
  return {target_checksum, point_checksum, distance_checksum};
}

std::tuple<double, std::int64_t, std::int64_t, std::int64_t>
benchmark_cvp(std::size_t dimension, int method) {
  const auto rows = cvp_basis(dimension);
  auto scaled_rows = rows;
  for (auto &entry : scaled_rows) {
    entry *= cvp_scale;
  }
  auto basis = to_matrix(scaled_rows, dimension);
  if (reduce(basis) != RED_SUCCESS) {
    throw std::runtime_error("fplll CVP preparation failed");
  }
  const auto integer_targets = make_targets(rows, dimension);
  std::vector<std::vector<Z_NR<mpz_t>>> targets;
  targets.reserve(targets_per_dimension);
  for (const auto &target : integer_targets) {
    targets.push_back(to_vector(target));
  }

  std::vector<std::vector<Z_NR<mpz_t>>> outputs(targets_per_dimension);
  for (std::size_t index = 0; index < targets.size(); ++index) {
    const int status =
        closest_vector(basis, targets[index], outputs[index], method);
    if (status != RED_SUCCESS) {
      throw std::runtime_error("fplll CVP warm-up failed");
    }
  }
  const auto [target_checksum, point_checksum, distance_checksum] =
      fingerprints(basis, targets, outputs);

  std::vector<std::chrono::nanoseconds> timings;
  timings.reserve(samples);
  for (std::size_t sample = 0; sample < samples; ++sample) {
    const auto start = std::chrono::steady_clock::now();
    for (std::size_t index = 0; index < targets.size(); ++index) {
      const int status =
          closest_vector(basis, targets[index], outputs[index], method);
      if (status != RED_SUCCESS) {
        throw std::runtime_error("fplll CVP failed");
      }
      status_sink = status;
    }
    timings.push_back(std::chrono::duration_cast<std::chrono::nanoseconds>(
        std::chrono::steady_clock::now() - start));
  }
  return {median_ns(std::move(timings), targets_per_dimension), target_checksum,
          point_checksum, distance_checksum};
}

} // namespace

int main() {
  try {
    std::cout << "library,operation,dimension,median_ns,target_fingerprint,"
                 "point_fingerprint,distance_fingerprint\n";
    std::cout << std::fixed << std::setprecision(2);
    for (const int dimension : dimensions) {
      const auto [lll_ns, basis_checksum] =
          benchmark_lll(static_cast<std::size_t>(dimension));
      std::cout << "fplll,lll," << dimension << ',' << lll_ns << ','
                << basis_checksum << ",0,0\n";
      const auto [fast_ns, fast_target, fast_point, fast_distance] =
          benchmark_cvp(static_cast<std::size_t>(dimension), CVPM_FAST);
      std::cout << "fplll,cvp_public_fast," << dimension << ',' << fast_ns
                << ',' << fast_target << ',' << fast_point << ','
                << fast_distance << '\n';
      const auto [proved_ns, proved_target, proved_point, proved_distance] =
          benchmark_cvp(static_cast<std::size_t>(dimension), CVPM_PROVED);
      std::cout << "fplll,cvp_public_proved," << dimension << ',' << proved_ns
                << ',' << proved_target << ',' << proved_point << ','
                << proved_distance << '\n';
    }
  } catch (const std::exception &error) {
    std::cerr << error.what() << '\n';
    return 1;
  }
  return status_sink;
}

File: benches/fplll_compare.rs
//! In-process comparison harness for the CVP operations shared with fplll.
//!
//! The fplll side of the comparison is `fplll_compare.cpp` in this directory;
//! build and run it separately at a pinned fplll version. This harness prints
//! the same CSV shape so the two outputs can be diffed directly.

use std::hint::black_box;
use std::time::{Duration, Instant};

use lattica::basis::Basis;
use lattica::reduce::{Delta, is_reduced};
use lattice_engine::{EnumerationScratch, Enumerator};

const DIMENSIONS: [usize; 3] = [8, 16, 24];
const TARGETS: usize = 128;
const SAMPLES: usize = 11;
const NODE_BUDGET: u64 = 1 << 30;
const CVP_SCALE: i128 = 1_009;

#[derive(Clone, Copy)]
struct Rng(u64);

impl Rng {
    fn next(&mut self) -> u64 {
        self.0 ^= self.0 << 13;
        self.0 ^= self.0 >> 7;
        self.0 ^= self.0 << 17;
        self.0
    }

    fn signed(&mut self, bound: i64) -> i64 {
        let width = u64::try_from(2 * bound + 1).unwrap();
        i64::try_from(self.next() % width).unwrap() - bound
    }
}

fn cvp_basis(dimension: usize) -> Vec<i128> {
    let mut basis = vec![0i128; dimension * dimension];
    for row in 0..dimension {
        basis[row * dimension + row] = 2;
        if row + 1 < dimension {
            basis[row * dimension + row + 1] = 1;
        }
    }
    basis
}

fn targets(basis: &[i128], dimension: usize) -> Vec<Vec<i128>> {
    let mut rng = Rng(0xa076_1d64_78bd_642f ^ u64::try_from(dimension).unwrap());
    let mut targets = Vec::with_capacity(TARGETS);
    for _ in 0..TARGETS {
        let lattice_coordinates: Vec<i64> = (0..dimension).map(|_| rng.signed(8)).collect();
        let mut target = vec![0i128; dimension];
        for (row, &coordinate) in lattice_coordinates.iter().enumerate() {
            for column in 0..dimension {
                target[column] +=
                    i128::from(coordinate) * basis[row * dimension + column] * CVP_SCALE;
            }
        }
        for entry in &mut target {
            *entry += i128::from(rng.signed(1_000));
        }
        targets.push(target);
    }
    targets
}

fn median(mut samples: Vec<Duration>) -> Duration {
    samples.sort_unstable();
    samples[samples.len() / 2]
}

fn per_operation(duration: Duration, operations: usize) -> f64 {
    duration.as_secs_f64() * 1e9 / f64::from(u32::try_from(operations).unwrap())
}

fn bounded_f64(value: i128) -> f64 {
    f64::from(i32::try_from(value).unwrap())
}

struct CvpCase {
    gram: lattica::Gram<i128>,
    reduced_basis: Basis<i128>,
    ambient_targets: Vec<Vec<i128>>,
    target_coordinates: Vec<Vec<f64>>,
}

fn prepare_cvp(dimension: usize) -> Result<CvpCase, Box<dyn std::error::Error>> {
    let rows = cvp_basis(dimension);
    let basis = Basis::from_rows(dimension, dimension, &rows)?;
    let gram = basis.gram()?;
    assert!(is_reduced(&gram, Delta::STRONG)?);
    let ambient_targets = targets(&rows, dimension);
    let mut target_coordinates = Vec::with_capacity(TARGETS);
    for target in &ambient_targets {
        let mut coordinate = vec![0.0; dimension];
        for column in 0..dimension {
            let mut value = bounded_f64(target[column]) / bounded_f64(CVP_SCALE);
            for (row, &coefficient) in coordinate.iter().take(column).enumerate() {
                value -= coefficient * bounded_f64(basis.as_matrix().get(row, column));
            }
            coordinate[column] = value / bounded_f64(basis.as_matrix().get(column, column));
        }
        target_coordinates.push(coordinate);
    }
    Ok(CvpCase {
        gram,
        reduced_basis: basis,
        ambient_targets,
        target_coordinates,
    })
}

fn cvp_fingerprints(case: &CvpCase, outputs: &[Vec<i64>]) -> (i128, i128, i128) {
    let mut target_checksum = 0i128;
    let mut point_checksum = 0i128;
    let mut distance_checksum = 0i128;
    let dimension = case.gram.dim();
    for (target_index, point) in outputs.iter().enumerate() {
        let mut ambient = vec![0i128; dimension];
        for (row, &coefficient) in point.iter().enumerate() {
            for (column, slot) in ambient.iter_mut().enumerate() {
                *slot += i128::from(coefficient) * case.reduced_basis.as_matrix().get(row, column);
            }
        }
        let mut distance = 0i128;
        for (column, value) in ambient.into_iter().enumerate() {
            let weight = i128::try_from(1 + target_index * dimension + column).unwrap();
            point_checksum += weight * value;
            target_checksum += weight * case.ambient_targets[target_index][column];
            let residual = case.ambient_targets[target_index][column] - value * CVP_SCALE;
            distance += residual * residual;
        }
        distance_checksum += i128::try_from(target_index + 1).unwrap() * distance;
    }
    (target_checksum, point_checksum, distance_checksum)
}

struct CvpMeasurement {
    cold_ns: f64,
    warm_ns: f64,
    target_fingerprint: i128,
    point_fingerprint: i128,
    distance_fingerprint: i128,
}

fn benchmark_cvp(dimension: usize) -> Result<CvpMeasurement, Box<dyn std::error::Error>> {
    let case = prepare_cvp(dimension)?;
    let enumerator = Enumerator::new(&case.gram)?;
    let mut scratch = EnumerationScratch::new();
    let mut outputs = vec![vec![0i64; dimension]; TARGETS];

    for (target, output) in case.target_coordinates.iter().zip(&mut outputs) {
        enumerator.nearest_ml(target, output, NODE_BUDGET, &mut scratch)?;
    }
    let (target_fingerprint, point_fingerprint, distance_fingerprint) =
        cvp_fingerprints(&case, &outputs);

    let mut warm_samples = Vec::with_capacity(SAMPLES);
    for _ in 0..SAMPLES {
        let start = Instant::now();
        for (target, output) in case.target_coordinates.iter().zip(&mut outputs) {
            enumerator.nearest_ml(target, output, NODE_BUDGET, &mut scratch)?;
            black_box(&output[0]);
        }
        warm_samples.push(start.elapsed());
    }

    let mut cold_samples = Vec::with_capacity(SAMPLES);
    for _ in 0..SAMPLES {
        let start = Instant::now();
        for (target, output) in case.target_coordinates.iter().zip(&mut outputs) {
            let cold_enumerator = Enumerator::new(&case.gram)?;
            let mut cold_scratch = EnumerationScratch::new();
            cold_enumerator.nearest_ml(target, output, NODE_BUDGET, &mut cold_scratch)?;
            black_box(&output[0]);
        }
        cold_samples.push(start.elapsed());
    }

    Ok(CvpMeasurement {
        cold_ns: per_operation(median(cold_samples), TARGETS),
        warm_ns: per_operation(median(warm_samples), TARGETS),
        target_fingerprint,
        point_fingerprint,
        distance_fingerprint,
    })
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!(
        "library,operation,dimension,median_ns,target_fingerprint,point_fingerprint,distance_fingerprint"
    );
    for dimension in DIMENSIONS {
        let cvp = benchmark_cvp(dimension)?;
        println!(
            "lattice-engine,cvp_cold,{dimension},{:.2},{},{},{}",
            cvp.cold_ns, cvp.target_fingerprint, cvp.point_fingerprint, cvp.distance_fingerprint
        );
        println!(
            "lattice-engine,cvp_warm,{dimension},{:.2},{},{},{}",
            cvp.warm_ns, cvp.target_fingerprint, cvp.point_fingerprint, cvp.distance_fingerprint
        );
    }
    Ok(())
}
